### PR TITLE
feat: honor SOURCE_DATE_EPOCH for chart archives

### LINKED ISSUE #31439: Add support for SOURCE_DATE_EPOCH in reproducible builds
### What would you like to be added?

See https://github.com/helm/helm/pull/31323#issuecomment-3333144185

### Why is this needed?

To enable someone to specify a date following the common environment variable name.

### PR DESCRIPTION
**Pull request by lohitkolluri**
Add `SOURCE_DATE_EPOCH` support (reproducible-builds standard) to produce deterministic `.tgz` archives across `helm package`, `install`, `upgrade`, `dep build`, and `dep update`.

### What changed

- **`pkg/chart/v2/chart.go`**: Added `StampModTimes(epoch time.Time)` method on Chart struct that recursively sets ModTime on chart, schema, lock, templates, raw files, and all dependency sub-charts.
- **`pkg/cmd/source_date_epoch.go`**: CLI-layer env var reader — calls `strconv.ParseInt` + `time.Unix` directly.
- **`pkg/action/package.go`, `pkg/downloader/manager.go`**: New `SourceDateEpoch *time.Time` fields; calls `ch.StampModTimes()` before archiving when set.
- **All 5 CLI commands**: `package`, `install`, `upgrade`, `dep build`, `dep update` read the env var and pass it through.
- **Removed**: Standalone utility functions and unused v3 duplicate (`internal/chart/v3/util/source_date_epoch.go`). The env is read in CLI layer only — library code stays pure.

### Test plan
- `go test ./pkg/cmd/ ./pkg/chart/v2/util/ ./internal/chart/v3/util/ ./pkg/action/ ./pkg/downloader/`
- New test verifies tar entry ModTime matches the epoch. Parsing tests cover set/unset/negative/boundary values.

Closes #31439

_AI disclosure: This PR was developed with assistance from Claude Code. All code was verified against Helm's contribution guidelines, DCO-signed, and tested._

### CODE REVIEW

File: pkg/action/package.go [hunk 1]
```diff
@@ -103,6 +103,15 @@ func (p *Package) Run(path string, _ map[string]any) (string, error) {
 		ch.Metadata.AppVersion = p.AppVersion
 	}
 
+	if epochStr, ok := os.LookupEnv("SOURCE_DATE_EPOCH"); ok && epochStr != "" {
```
Comment (TerryHowe) [inline pkg/action/package.go]: The env-var read belongs in pkg/cmd/package.go, and action.Package should expose a SourceDateEpoch
  *time.Time field that the CLI populates.

File: pkg/downloader/manager.go [hunk 2]
```diff
@@ -888,6 +888,15 @@ func tarFromLocalDir(chartpath, name, repo, version, destPath string) (string, e
 		return "", err
 	}
 
+	if epochStr, ok := os.LookupEnv("SOURCE_DATE_EPOCH"); ok && epochStr != "" {
```
Comment (TerryHowe) [inline pkg/downloader/manager.go]: Same kind of thing here about the env read.

File: pkg/action/package.go [hunk 3]
```diff
@@ -103,6 +103,15 @@ func (p *Package) Run(path string, _ map[string]any) (string, error) {
...
+		epoch, err := chartutil.ParseSourceDateEpochValue(epochStr)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "warning: invalid SOURCE_DATE_EPOCH: %v (ignoring)\n", err)
```
Comment (TerryHowe) [inline pkg/action/package.go]: When ytou move the env read, maybe move the parse as well and make it fatal.

**Review by TerryHowe — CHANGES_REQUESTED** [review]

**Comment by TerryHowe** [comment]
Conflicts, needs a rebase please

**Comment by lohitkolluri** [comment]
Just wanted to run Ponytail to check for any redundant or unused changes. Sorry for the delay.

File: pkg/cmd/install.go [hunk 4]
```diff
@@ -296,6 +296,11 @@ func runInstall(args []string, client *action.Install, valueOpts *values.Options
 		slog.Warn("this chart is deprecated")
 	}
 
+	sourceDateEpoch, err := sourceDateEpochFromEnv()
+	if err != nil {
+		return nil, err
+	}
```
Comment (TerryHowe) [inline pkg/cmd/install.go]: This check should probably be done inside `if req := ac.MetaDependencies(); len(req) > 0 {` because that if would indicate there is something to build.

**Review by TerryHowe — CHANGES_REQUESTED** [review]

File: pkg/cmd/install.go (hunk 4 above)
Comment (lohitkolluri) [reply pkg/cmd/install.go]: Done, moved it inside that block in install.go. Same fix applied to upgrade.go too. Thanks.

**Review by TerryHowe — CHANGES_REQUESTED** [review]
Lint failure
```
  Running [/home/runner/golangci-lint-2.11.3-linux-amd64/golangci-lint run] in [/home/runner/work/helm/helm] ...
  Error: pkg/cmd/source_date_epoch.go:37:15: unnecessary-format: unnecessary use of formatting function "fmt.Errorf", you can replace it with "errors.New" (revive)
  		return nil, fmt.Errorf("invalid SOURCE_DATE_EPOCH: must not be negative")
  		            ^
  1 issues:
```

**Comment by lohitkolluri** [comment]
CI is green. @TerryHowe anything else needed on my end?

**Review by TerryHowe — DISMISSED** [review]
/lgtm

**Comment by lohitkolluri** [comment]
@TerryHowe thanks for the /lgtm. I see the "docs needed" label  should I include the docs in this PR itself, or raise a separate follow-up PR for documentation?

**Comment by TerryHowe** [comment]
> @TerryHowe thanks for the /lgtm. I see the "docs needed" label should I include the docs in this PR itself, or raise a separate follow-up PR for documentation?

That was a note to create docs in helm-www, but some short docs in this PR would be a good addition. Thanks.

**Comment by lohitkolluri** [comment]
Added short docs in d8979b1 — documented `$SOURCE_DATE_EPOCH` in the env var table in `helm help`. Full docs still belongs in helm-www. PTAL.

**Review by TerryHowe — DISMISSED** [review]

**Comment by lohitkolluri** [comment]
Just following up, who handles the second review here? Is Copilot's review enough or does a maintainer need to approve?

**Comment by TerryHowe** [comment]
> Just following up, who handles the second review here? Is Copilot's review enough or does a maintainer need to approve?

Maintainer. It might be a while.

**Comment by lohitkolluri** [comment]
No worries its my first time contributing to Helm so i was just curious on the workflow here.

**Review by gjenkins8 — COMMENTED** [review]
Thanks for the clean implementation. Changes LGTM, can you please rebase/merge main to fix conflicts. And I can approve.

**Comment by lohitkolluri** [comment]
@gjenkins8 Thanks! I've resolved the merge conflicts and pushed the updated branch. PTAL when you have a chance.

**Review by gjenkins8 — APPROVED** [review]

**Review by TerryHowe — APPROVED** [review]
/lgtm

**Comment by promptless-for-oss** [comment]
Promptless prepared a documentation update related to this change.

Triggered by [helm/helm#32162](https://github.com/helm/helm/pull/32162) (thanks @lohitkolluri)

Added a new "Build Reproducible Chart Archives" section to the Charts Tips & Tricks how-to guide documenting `SOURCE_DATE_EPOCH` support: the headline `helm package` use case and `git log -1 --pretty=%ct` idiom, value semantics (non-negative whole seconds in UTC; invalid values fail), and the scope caveat that `install`/`upgrade`/`dependency build`/`dependency update` honor it only when re-archiving local `file://` dependencies.

Review: [Document SOURCE_DATE_EPOCH for reproducible chart archives](https://github.com/helm/helm-www/pull/2180)
